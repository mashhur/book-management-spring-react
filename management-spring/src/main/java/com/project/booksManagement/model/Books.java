package com.project.booksManagement.model;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collection;

import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.Lob;
import javax.persistence.ManyToMany;
import javax.persistence.OneToMany;
import javax.persistence.OneToOne;
import javax.persistence.Transient;


@Entity
public class Books implements Serializable{
	
	
	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	int bid;
	@Column
	String title;
	@Column
	String author;
	@Column
	double price;
	@Column
	int year;
	@Column
	String publisher;
	
    @Lob
	String cover;
	
	@Lob
	String description;
	
	@OneToMany(fetch = FetchType.LAZY)
    Collection<Ratings> ratings;
	
	@Transient
	String categoryName;
	
    @OneToOne(cascade = CascadeType.PERSIST)
    @JoinColumn(name = "category_name", referencedColumnName = "catid")
	Category category;

	@ManyToMany(fetch = FetchType.LAZY)  
    Collection<BookOrder> bookOrderId = new ArrayList<>();
	
	
	public Books(int bid, String title, String author, double price, int year, String publisher, String cover,
			String description, Category category) {
		super();
		this.bid = bid;
		this.title = title;
		this.author = author;
		this.price = price;
		this.year = year;
		this.publisher = publisher;
		this.cover = cover;
		this.description = description;
		this.category = category;
	}
	
	public Books(int bid, String title, String author, double price, int year, String publisher, String cover,
			String description, String categoryName) {
		super();
		this.bid = bid;
		this.title = title;
		this.author = author;
		this.price = price;
		this.year = year;
		this.publisher = publisher;
		this.cover = cover;
		this.description = description;
		this.categoryName = categoryName;
	}


	
	

	public Collection<Ratings> getRatings() {
		return ratings;
	}

	public void setRatings(Collection<Ratings> ratings) {
		this.ratings = ratings;
	}

	public Collection<BookOrder> getBookOrderId() {
		return bookOrderId;
	}

	public void setBookOrderId(Collection<BookOrder> bookOrderId) {
		this.bookOrderId = bookOrderId;
	}

	public Books() {
		// TODO Auto-generated constructor stub
	}


	public int getBid() {
		return bid;
	}


	public void setBid(int bid) {
		this.bid = bid;
	}


	public String getTitle() {
		return title;
	}


	public void setTitle(String title) {
		this.title = title;
	}


	public String getAuthor() {
		return author;
	}


	public void setAuthor(String author) {
		this.author = author;
	}


	public double getPrice() {
		return price;
	}


	public void setPrice(double price) {
		this.price = price;
	}


	public int getYear() {
		return year;
	}


	public void setYear(int year) {
		this.year = year;
	}


	public String getPublisher() {
		return publisher;
	}


	public void setPublisher(String publisher) {
		this.publisher = publisher;
	}


	public String getCover() {
		return cover;
	}


	public void setCover(String cover) {
		this.cover = cover;
	}


	public String getDescription() {
		return description;
	}


	public void setDescription(String description) {
		this.description = description;
	}


	public Category getCategory() {
		return category;
	}


	public void setCategory(Category category) {
		this.category = category;
	}


	public String getCategoryName() {
		return categoryName;
	}
	
	public void setCategoryName(String categoryName) {
		this.categoryName = categoryName;
	}
	
	
}
