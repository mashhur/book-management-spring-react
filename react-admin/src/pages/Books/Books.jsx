import { filter } from "lodash";
import { Icon } from "@iconify/react";
import { sentenceCase } from "change-case";
import { useState, useEffect } from "react";
import plusFill from "@iconify/icons-eva/plus-fill";
import {
	Link as RouterLink,
	useNavigate,
} from "react-router-dom";
// material
import {
	Card,
	Table,
	Stack,
	Avatar,
	Button,
	Checkbox,
	TableRow,
	TableBody,
	TableCell,
	Container,
	Typography,
	TableContainer,
	TablePagination,
} from "@mui/material";
// components
import Page from "../../components/Page";
import Label from "../../components/Label";
import Scrollbar from "../../components/Scrollbar";
import SearchNotFound from "../../components/SearchNotFound";
import {
	UserListHead,
	UserListToolbar,
	UserMoreMenu,
} from "../../components/_dashboard/user";
//
import {
	getAllBooks,
	deleteBook,
} from "../../service/BooksService";

const TABLE_HEAD = [
	{ id: "id", label: "Id", alignRight: false },
	{ id: "title", label: "Title", alignRight: false },
	{ id: "author", label: "Author", alignRight: false },
	{
		id: "publisher",
		label: "Publisher",
		alignRight: false,
	},
	{ id: "year", label: "Year", alignRight: false },
	{ id: "price", label: "Price", alignRight: false },
	{ id: "category", label: "Category", alignRight: false },
];

export default function Books() {
	const [page, setPage] = useState(0);
	const [USERLIST, setUSERLIST] = useState([]);
	const [rowsPerPage, setRowsPerPage] = useState(5);

	const navigate = useNavigate();

	useEffect(() => {
		getAllBooks().then((res) => {
			console.log(res.data);
			setUSERLIST(res.data);
		});
	}, []);

	const deleteHandler = (id) => {
		deleteBook(id).then(() => {
			const newList = filter(
				USERLIST,
				(user) => user.bid !== id
			);
			setUSERLIST(newList);
		});
	};

	const handleRequestSort = (event, property) => {};
	const handleSelectAllClick = (event) => {};
	const handleClick = (event, name) => {};

	const handleChangePage = (event, newPage) => {
		setPage(newPage);
	};

	const handleChangeRowsPerPage = (event) => {
		setRowsPerPage(parseInt(event.target.value, 10));
		setPage(0);
	};

	const handleFilterByName = (event) => {};

	const filteredUsers = USERLIST;

	const isUserNotFound = filteredUsers.length === 0;

	return (
		<Page title="Admin">
			<Container>
				<Stack
					direction="row"
					alignItems="center"
					justifyContent="space-between"
					mb={5}>
					<Typography variant="h4" gutterBottom>
						Books
					</Typography>
					<Button
						variant="contained"
						onClick={() => {
							navigate("add");
						}}
						startIcon={<Icon icon={plusFill} />}>
						Add a book
					</Button>
				</Stack>

				<Card>
					{/* <UserListToolbar
            numSelected={selected.length}
            filterName={filterName}
            onFilterName={handleFilterByName}
          /> */}

					<Scrollbar>
						<TableContainer sx={{ minWidth: 800 }}>
							<Table>
								<UserListHead
									order={null}
									orderBy={null}
									headLabel={TABLE_HEAD}
									rowCount={USERLIST.length}
									numSelected={0}
									onRequestSort={handleRequestSort}
									onSelectAllClick={handleSelectAllClick}
								/>
								<TableBody>
									{filteredUsers
										.slice(
											page * rowsPerPage,
											page * rowsPerPage + rowsPerPage
										)
										.map((row) => {
											const {
												bid,
												title,
												author,
												year,
												price,
												category,
												cover,
												description,
												publisher,
											} = row;

											const isItemSelected = false;

											return (
												<TableRow
													hover
													key={bid}
													tabIndex={-1}
													role="checkbox"
													selected={false}
													aria-checked={false}>
													<TableCell padding="checkbox">
														<Checkbox
															checked={false}
															onChange={() => {}}
														/>
													</TableCell>
													<TableCell
														component="th"
														scope="row"
														padding="none">
														<Stack
															direction="row"
															alignItems="center"
															spacing={2}>
															<Typography
																variant="subtitle2"
																noWrap>
																{bid}
															</Typography>
														</Stack>
													</TableCell>
													<TableCell align="left">
														{title}
													</TableCell>

													<TableCell align="left">
														{author}
													</TableCell>
													<TableCell align="left">
														{publisher}
													</TableCell>
													<TableCell align="left">
														{year}
													</TableCell>

													<TableCell align="left">
														{price}
													</TableCell>
													<TableCell align="left">
														{category.name}
													</TableCell>

													<TableCell align="right">
														<UserMoreMenu
															deleteHandler={deleteHandler}
															id={bid}
														/>
													</TableCell>
												</TableRow>
											);
										})}
								</TableBody>
							</Table>
						</TableContainer>
					</Scrollbar>

					<TablePagination
						rowsPerPageOptions={[5, 10, 25]}
						component="div"
						count={USERLIST.length}
						rowsPerPage={rowsPerPage}
						page={page}
						onPageChange={handleChangePage}
						onRowsPerPageChange={handleChangeRowsPerPage}
					/>
				</Card>
			</Container>
		</Page>
	);
}
