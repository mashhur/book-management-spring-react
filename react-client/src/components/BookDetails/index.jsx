import React from "react";
import { useDispatch } from "react-redux";
import { useNavigate } from "react-router-dom";
import {
	addBook,
	addBooktoCart,
} from "../../store/slices/userSlice";
import { TiShoppingCart } from "react-icons/ti";
import { AiFillStar } from "react-icons/ai";

function BookDetails({ book }) {
	const dispatch = useDispatch();
	const navigate = useNavigate();
	const buyBook = () => {
		dispatch(addBook(book));
		navigate("/checkout");
	};

	const addToCart = () => {
		dispatch(addBooktoCart(book));
	};

	const getBookRatings = (book) => {
		let ratings = 0;
		if (book.ratings) {
			ratings = book.ratings.reduce((acc, curr) => {
				return acc + curr.rating;
			}, 0);
			ratings = (ratings / book.ratings.length) % 5;
		}
		return ratings || 0;
	};

	return (
		<div
			style={{
				borderTopLeftRadius: "1.4rem",
				borderBottomLeftRadius: "1.4rem",
			}}
			className="offcanvas ps-2  offcanvas-end"
			tabindex="-1"
			id="offcanvasExample"
			aria-labelledby="offcanvasExampleLabel">
			{/* <div className="offcanvas-header">
				<h5
					className="offcanvas-title text-capitalize"
					id="offcanvasExampleLabel">
					{book.title}
				</h5>
				<button
					type="button"
					className="btn-close text-reset"
					data-bs-dismiss="offcanvas"
					aria-label="Close"></button>
			</div> */}
			<div className="offcanvas-body text-center ">
				<div className="  my-3 p-3">
					<img
						style={{
							height: "320px",
							borderRadius: "1.3rem",
							transform: "rotate(-0.02turn)",
						}}
						className="img-fluid "
						src={book.cover}
						alt="Card image cap"></img>
				</div>
				<h3 className="mt-3 fs-3">{book.title}</h3>

				<p
					className="  text-muted "
					id=""
					style={{ lineHeight: "1.2", fontSize: "14px" }}>
					{book.description}
				</p>

				{/* add here */}
				<div className="mt-3">
					<div className="d-flex justify-content-start align-items-center">
						<small>AUTHOR </small>
						<div
							style={{
								borderRadius: "2rem",
								opacity: "0.8",
							}}
							className="alert-success font-weight-light ms-2  py-1 px-3 ">
							<small>
								<b>{book.author}</b>
							</small>
						</div>
					</div>
					<div className="d-flex justify-content-start mt-2 align-items-center">
						<small>PUBLISHER</small>
						<div
							style={{
								borderRadius: "2rem",
								fontSize: "14px",
								maxLines: "1",
								overflow: "hidden",
								textOverflow: "ellipsis",
								height: "1.9rem",
								opacity: "0.8",
							}}
							className="alert-success font-weight-light ms-2  py-1 px-2 ">
							<b>{book.publisher}</b>
						</div>
					</div>

					<div className="d-flex justify-content-start mt-2 align-items-center">
						<small>YEAR </small>
						<div
							style={{
								borderRadius: "2rem",
								opacity: "0.8",
							}}
							className="alert-success font-weight-light ms-2  py-1 px-3 ">
							<small>
								<b>{book.year}</b>
							</small>
						</div>
					</div>
					<div className="d-flex justify-content-start mt-2 align-items-center">
						<AiFillStar
							style={{
								color: "rgb(249, 210, 50)",
								fontSize: "30px",
							}}
						/>
						{getBookRatings(book)}
					</div>

					<div className="mt-3 d-flex w-100">
						<button
							onClick={buyBook}
							style={{
								borderRadius: "1rem",
								fontWeight: "bold",
								fontSize: "1.0rem",
								padding: "0.9rem 1rem",
							}}
							type="button"
							className="btn me-2 flex-fill btn-dark  hover-btn">
							Buy Now at â‚¹{book.price}
						</button>
						<div
							data-bs-dismiss="offcanvas"
							onClick={addToCart}
							style={{
								borderRadius: "1rem",
								fontWeight: "bold",
								fontSize: "1.0rem",
								padding: "0.9rem 1rem",
							}}
							className="btn btn-success hover-btn">
							<TiShoppingCart
								style={{
									color: "#fff",
									fontWeight: "bold",
									fontSize: "2rem",
								}}
							/>
						</div>
					</div>
				</div>
			</div>
		</div>
	);
}

export default BookDetails;
