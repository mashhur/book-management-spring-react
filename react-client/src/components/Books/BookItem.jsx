import PropTypes from "prop-types";
import { AiFillStar } from "react-icons/ai";

export const getBookRatings = (book) => {
	let ratings = 0;
	if (book.ratings) {
		ratings = book.ratings.reduce((acc, curr) => {
			return acc + curr.rating;
		}, 0);
		ratings = (ratings / book.ratings.length) % 5;
	}
	return ratings || 0;
};

function BookItem({ book, setSelectedBook }) {
	const handleClick = () => {
		setSelectedBook(book);
	};

	return (
		<>
			<div
				data-test="book-item"
				onClick={handleClick}
				style={{ cursor: "pointer" }}
				className="col book-item-hover "
				data-bs-toggle="offcanvas"
				data-bs-target="#offcanvasExample"
				aria-controls="offcanvasExample">
				<div className="p-2 ">
					<img
						alt={book.title}
						style={{
							width: "88%",
							height: "220px",
							objectFit: "cover",
							borderRadius: "1.2rem",
						}}
						src={book.cover}
						className=" "
					/>
					<div className="py-2 pe-2 d-flex justify-content-between align-items-center ">
						<div className="">
							<h6 className="card-title mb-0">
								{book.title}
							</h6>
							<small
								style={{ fontSize: "12px" }}
								className="text-muted ">
								{book.author}
							</small>
						</div>
						<small
							style={{ fontSize: "12px" }}
							className="text-muted d-flex  me-3">
							<AiFillStar
								style={{
									color: "rgb(249, 210, 50)",
									fontSize: "18px",
								}}
							/>
							{getBookRatings(book)}
						</small>
					</div>
				</div>
			</div>
		</>
	);
}

BookItem.propTypes = {
	book: PropTypes.object.isRequired,
	setSelectedBook: PropTypes.func.isRequired,
};

export default BookItem;
