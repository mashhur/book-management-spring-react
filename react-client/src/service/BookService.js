const axios = require("axios");

const BASE_URL = "http://localhost:8080/api/books";

class BookService {
	async getAllBooks() {
		return axios.get(BASE_URL);
	}
	async getBooksByCategory(category) {
		return axios.get(BASE_URL + "/c/" + category);
	}
}

module.exports = new BookService();

// [
// 	{
// 	  "id": 2,
// 	  "title": "The Hobbit",
// 	  "author": null,
// 	  "price": 100.0,
// 	  "year": 2022,
// 	  "": "Allen & Unwin",
// 	  "cover": null,
// 	  "description": null,
// 	  "categoryName": null,
// 	  "category": null
// 	}
//   ]
