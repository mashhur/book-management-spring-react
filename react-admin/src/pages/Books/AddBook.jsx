import * as Yup from "yup";
import { useState, useEffect } from "react";
import {
	Link as RouterLink,
	useNavigate,
} from "react-router-dom";
import {
	Card,
	Container,
	Stack,
	Typography,
	InputLabel,
	Input,
	FormHelperText,
	FormControl,
	TextField,
	Select,
	MenuItem,
} from "@mui/material";
import { LoadingButton } from "@mui/lab";
import { useFormik, Form, FormikProvider } from "formik";

import Page from "../../components/Page";
import Scrollbar from "../../components/Scrollbar";

import { getAllCategories } from "../../service/CategoryService";
import { createBook } from "../../service/BooksService";

export default function AddBooks() {
	const navigate = useNavigate();
	const [category, setCategory] = useState("");

	useEffect(() => {}, []);

	const RegisterSchema = Yup.object().shape({
		title: Yup.string()
			.min(2, "Too Short!")
			.max(50, "Too Long!")
			.required("Title required"),
		author: Yup.string()
			.min(2, "Too Short!")
			.max(50, "Too Long!")
			.required("Author required"),
		publisher: Yup.string()
			.min(2, "Too Short!")
			.max(50, "Too Long!")
			.required("Publisher required"),
		year: Yup.number()
			.test(
				"len",
				"Must be exactly 4 characters",
				(val) =>
					!val || (val && val.toString().length === 4)
			)
			.required("Year required"),
		price: Yup.number().required("Price required"),
		description: Yup.string()
			.min(2, "Too Short!")
			.max(500, "Too Long!")
			.required("Description required"),
		cover: Yup.string().required("Cover image required"),
	});

	const formik = useFormik({
		initialValues: {
			title: "",
			author: "",
			publisher: "",
			year: "",
			price: "",
			description: "",
			cover: "",
		},
		validationSchema: RegisterSchema,
		onSubmit: (values) => {
			values.categoryName = category;
			console.log(values);
			createBook(values).then((_res) => {
				window.history.back();
			});
		},
	});

	const {
		errors,
		touched,
		handleSubmit,
		isSubmitting,
		getFieldProps,
	} = formik;

	return (
		<Page title="Admin">
			<Scrollbar>
				<Container>
					<Stack
						direction="row"
						alignItems="center"
						justifyContent="space-between"
						mb={2}>
						<Typography variant="h4"> Add Book</Typography>
					</Stack>
					<FormikProvider value={formik}>
						<Form
							autoComplete="off"
							noValidate
							onSubmit={handleSubmit}>
							<Stack
								justifyContent="center"
								sx={{ width: "55%" }}
								spacing={3}>
								<TextField
									fullWidth
									label="Title"
									{...getFieldProps("title")}
									error={Boolean(
										touched.title && errors.title
									)}
									helperText={touched.title && errors.title}
								/>

								<TextField
									fullWidth
									label="Author"
									{...getFieldProps("author")}
									error={Boolean(
										touched.author && errors.author
									)}
									helperText={
										touched.author && errors.author
									}
								/>

								<Stack direction="row" spacing={3}>
									<TextField
										fullWidth
										label="Publisher"
										{...getFieldProps("publisher")}
										error={Boolean(
											touched.publisher && errors.publisher
										)}
										helperText={
											touched.publisher && errors.publisher
										}
									/>
									<TextField
										fullWidth
										label="Year"
										{...getFieldProps("year")}
										error={Boolean(
											touched.year && errors.year
										)}
										helperText={touched.year && errors.year}
									/>
								</Stack>
								<Stack direction="row" spacing={3}>
									<SelectComponent
										category={category}
										setCategory={setCategory}
									/>

									<TextField
										fullWidth
										label="Price"
										{...getFieldProps("price")}
										error={Boolean(
											touched.price && errors.price
										)}
										helperText={
											touched.price && errors.price
										}
									/>
								</Stack>
								<TextField
									fullWidth
									label="Cover"
									{...getFieldProps("cover")}
									error={Boolean(
										touched.cover && errors.cover
									)}
									helperText={touched.cover && errors.cover}
								/>
								<TextField
									fullWidth
									multiline
									id="outlined-multiline-static"
									rows={3}
									label="Description"
									{...getFieldProps("description")}
									error={Boolean(
										touched.description &&
											errors.description
									)}
									helperText={
										touched.description &&
										errors.description
									}
								/>

								<LoadingButton
									fullWidth
									size="large"
									type="submit"
									variant="contained"
									loading={isSubmitting}>
									Add book
								</LoadingButton>
							</Stack>
						</Form>
					</FormikProvider>
				</Container>
			</Scrollbar>
		</Page>
	);
}

function SelectComponent({ category, setCategory }) {
	const [categoryData, setCategoryData] = useState([]);

	const handleChange = (event) => {
		setCategory(event.target.value);
	};

	useEffect(() => {
		getAllCategories().then((_category) => {
			setCategoryData(_category.data);
		});
	}, []);

	return (
		<FormControl fullWidth>
			<InputLabel id="demo-simple-select-label">
				Category
			</InputLabel>
			<Select
				labelId="demo-simple-select-label"
				id="demo-simple-select"
				value={category}
				label="Category"
				onChange={handleChange}>
				{categoryData.map((item) => (
					<MenuItem key={item.id} value={item.name}>
						{item.name}
					</MenuItem>
				))}
			</Select>
		</FormControl>
	);
}
