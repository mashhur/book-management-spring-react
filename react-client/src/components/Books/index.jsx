import React from "react";
import BookService from "../../service/BookService";
import BookDetails from "../BookDetails";
import BookItem from "./BookItem";
import PropTypes from "prop-types";

function Books({ category }) {
	const [bookState, setBookState] = React.useState({
		books: [],
		selectedBook: {
			title: null,
		},
		error: null,
		isLoading: true,
	});

	const { error, isLoading, books, selectedBook } =
		bookState;

	const setbooks = (books) => {
		setBookState({
			...bookState,
			books: books,
			isLoading: false,
		});
	};

	const setSelectedBook = (selectedBook) => {
		setBookState({
			...bookState,
			selectedBook: selectedBook,
		});
	};

	const setError = (error) => {
		setBookState({
			...bookState,
			isLoading: false,
			error: error,
		});
	};
	const chooseSelectedBookTitle = (title) => {
		setSelectedBook(title);
	};

	React.useEffect(() => {
		if (category === "all") {
			BookService.getAllBooks()
				.then((res) => {
					setbooks(res.data);
				})
				.catch((err) => {
					setError(err.message);
				});
		} else {
			BookService.getBooksByCategory(category)
				.then((res) => {
					setbooks(res.data);
				})
				.catch((err) => {
					setError(err.message);
				});
		}
	}, [category]);

	return (
		<div data-test="component-books">
			{isLoading ? (
				<div
					data-test="loading-spinner"
					className="text-center">
					<div
						className="spinner-border text-primary"
						role="status"></div>
				</div>
			) : error ? (
				<div
					data-test="error-message"
					className="text-center">
					<div className="alert text-danger">{error}</div>
				</div>
			) : (
				<div data-test="books-list">
					{selectedBook.title !== null ? (
						<BookDetails book={selectedBook} />
					) : (
						""
					)}
					<div className="row row-cols-2 row-cols-md-5 mt-4 ms-3 g-2">
						{books.map((book) => (
							<BookItem
								setSelectedBook={chooseSelectedBookTitle}
								book={book}
								key={book.bid}
							/>
						))}
					</div>
				</div>
			)}
		</div>
	);
}

Books.propTypes = {
	category: PropTypes.string.isRequired,
};

export default Books;
