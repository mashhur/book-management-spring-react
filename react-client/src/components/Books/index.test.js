import React from "react";
import { mount } from "enzyme";
import moxios from "moxios";
import Books from ".";
import {
	checkProp,
	findByTestAttr,
} from "../../tests/testUtils";
import BookItem from "./BookItem";
import BookDetails from "../BookDetails";
import BookService from "../../service/BookService";
import { DummyData } from "../../tests/dummyData";

const books = DummyData.bookArray;

const initialState = {
	books: [],
	selectedBook: {
		title: null,
	},
	error: null,
	isLoading: true,
};

const defaultProps = {
	category: "all",
};

const mockSetCurrentGuess = jest.fn();
jest.mock("react", () => ({
	...jest.requireActual("react"),
	useState: (i) => [initialState, mockSetCurrentGuess],
}));

const setup = (props = {}) => {
	const setupProps = { ...defaultProps, ...props };
	return mount(<Books {...setupProps} />);
};

describe("Testing books component", () => {
	beforeAll(() => {
		moxios.install();
	});
	afterAll(() => {
		moxios.uninstall();
	});

	it("renders without any error", async () => {
		const wrapper = setup();
		const component = await findByTestAttr(
			wrapper,
			"component-books"
		);
		expect(component.length).toBe(1);
	});

	it("should not renders `BookDetails` until selected", async () => {
		const wrapper = setup();
		expect(wrapper.find(BookDetails).exists()).toBe(false);
	});

	it("renders without any error with required props", () => {
		checkProp(Books, defaultProps);
	});

	describe("when `getAllBooks()` is called", () => {
		it("return the correct repsonse", async () => {
			await moxios.wait(() => {
				const req = moxios.requests.mostRecent();
				req.respondWith({
					status: 200,
					response: books,
				});
			});
			return BookService.getAllBooks().then((respsonse) => {
				expect(respsonse.data).toEqual(books);
			});
		});
		//response error has been tested.
		it("return error response", async () => {
			await moxios.wait(() => {
				const req = moxios.requests.mostRecent();
				const errResp = {
					status: 500,
					response: {
						msg: "The server is unable to respond",
					},
				};
				req.reject(errResp);
			});

			return BookService.getAllBooks().catch((error) => {
				// eslint-disable-next-line jest/no-conditional-expect
				expect(error.status).toEqual(500);
			});
		});
	});

	it("`all` is passed should call the getAllBooks()", async () => {
		jest
			.spyOn(React, "useEffect")
			.mockImplementationOnce((f) => f());
		jest.spyOn(BookService, "getAllBooks");
		setup();
		expect(BookService.getAllBooks).toHaveBeenCalledTimes(
			1
		);
	});

	describe("when api call is made", () => {
		describe("should render loading response", () => {
			let wrapper;
			beforeEach(() => {
				React.useState = jest.fn(() => [
					initialState,
					jest.fn(),
				]);

				wrapper = setup();
			});
			it("should render loading", async () => {
				const component = await findByTestAttr(
					wrapper,
					"loading-spinner"
				);
				expect(component.length).toBe(1);
			});

			it("should match the default loading snapshot", () => {
				expect(wrapper).toMatchSnapshot();
			});
		});

		//testing error state
		describe("when request fails", () => {
			let wrapper;
			beforeEach(() => {
				const initalValue = {
					...initialState,
					isLoading: false,
					error: "An error has occured",
				};

				React.useState = jest.fn(() => [
					initalValue,
					jest.fn(),
				]);
				wrapper = setup();
			});
			it("should render error msg", async () => {
				const wrapper = setup();
				const component = await findByTestAttr(
					wrapper,
					"error-message"
				);
				expect(component.length).toBe(1);
			});

			it("should match the error snapshot", () => {
				expect(wrapper).toMatchSnapshot();
			});
		});

		//testing categories state
		describe("when request is success", () => {
			let wrapper;
			beforeEach(() => {
				const initalValue = {
					...initialState,
					isLoading: false,
					books: books,
				};

				React.useState = jest.fn(() => [
					initalValue,
					jest.fn(),
				]);

				wrapper = setup();
			});
			it("should render items", async () => {
				expect(wrapper.find(BookItem).length).toBe(1);
			});

			it("should match the success snapshot", () => {
				expect(wrapper).toMatchSnapshot();
			});
		});
	});
});
