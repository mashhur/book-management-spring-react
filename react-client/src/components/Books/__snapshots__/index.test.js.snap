// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing books component when an api call is made while loading should match the default loading snapshot 1`] = `ReactWrapper {}`;

exports[`Testing books component when api call is made should render loading response should match the default loading snapshot 1`] = `ReactWrapper {}`;

exports[`Testing books component when api call is made when request fails should match the error snapshot 1`] = `ReactWrapper {}`;

exports[`Testing books component when api call is made when request is success should match the success snapshot 1`] = `ReactWrapper {}`;
